mountpoint-s3/src/prefetch.rs

```diff
@@ -284,31 +323,47 @@ where
         }
         assert_eq!(self.next_sequential_read_offset, offset);
 
-        self.prepare_requests();
+        if self.backpressure_task.is_none() {
+            self.backpressure_task = Some(self.spawn_read_backpressure_request());
+        }
 
         let mut response = ChecksummedBytes::default();
         while to_read > 0 {
-            let Some(current_task) = self.current_task.as_mut() else {
-                // If [prepare_requests] didn't spawn a request, we've reached the end of the object.
+            let Some(current_task) = self.backpressure_task.as_mut() else {
                 trace!(offset, length, "read beyond object size");
                 break;
             };
             debug_assert!(current_task.remaining() > 0);
 
+            if let Some(len) = self.next_read_window_size {
+                trace!(
+                    current_window = self.current_read_window_size,
+                    next_window = len,
+                    "incrementing read window"
+                );
+                current_task.increment_read_window(len);
+                self.current_read_window_size = len;
+                self.next_read_window_size = None;
+            }
             let part = match current_task.read(to_read as usize).await {
```

Comment:
not clear how do we make sure that this will never block (which would mean a deadlock as this function is the only one increasing the window), I believe you guarantee that `to_read` is always in the window, a comment may help